bugwarrior/services/phab.py

```diff
@@ -57,6 +61,46 @@ def get_default_description(self):
             cls=self.extra['type'],
         )
 
+    def _normalize_to_tag(self, s):
+        return re.sub(r'[^a-zA-Z0-9]', '_', s)
+
+    def get_tags(self):
+        tags = []
+
+        if not self.origin['should_set_tags']:
+            return tags
+
+        projects = self.extra['projects']
+
+        template = Template(self.origin['tag_template'])
+        for p in projects:
+            slug = p['fields']['slug']
+            slug = self.apply_tag_substitutions(slug)
+            context = self.record.copy()
+            context.update({
+                'project_slug': self._normalize_to_tag(slug),
+            })
+            rendered = template.render(context)
+            tags.append(rendered)
+        return tags
+
+    def apply_tag_substitutions(self, tag):
+        """Applies a user's regexes and replacement strings to a tag
+
+        :self: Gets from self a list of substitutions with form ['find#replace', 'findme#replacewithme', ...]
+        :returns: A function that takes a tag and returns another tag
+        """
+        exprs = self.origin['tag_substitutions']
+        if len(exprs) == 0:
+            return tag
+        exprs = [e.split('#', 1) for e in exprs]
+
+        def reducer(acc, upd):
+            [pattern, sub] = upd
+            return re.sub(pattern, sub, acc)
+
+        return reduce(reducer, exprs, tag)
```

Comment:
This will need to be changed for python3. You could `try` to `import reduce from functools` or you could switch to a for-loop [as suggested by the python maintainers](https://docs.python.org/3.0/whatsnew/3.0.html#builtins). Your call.